core/src/main/java/io/substrait/expression/proto/ProtoExpressionConverter.java

```diff
@@ -109,7 +111,54 @@ public Expression from(io.substrait.proto.Expression expr) {
         yield ImmutableExpression.ScalarFunctionInvocation.builder()
             .addAllArguments(args)
             .declaration(declaration)
-            .outputType(protoTypeConverter.from(expr.getScalarFunction().getOutputType()))
+            .outputType(protoTypeConverter.from(scalarFunction.getOutputType()))
+            .build();
+      }
+      case WINDOW_FUNCTION -> {
+        var windowFunction = expr.getWindowFunction();
+        var functionReference = windowFunction.getFunctionReference();
+        var declaration = lookup.getWindowFunction(functionReference, extensions);
+        var pF = new FunctionArg.ProtoFrom(this, protoTypeConverter);
+        var args =
+            IntStream.range(0, windowFunction.getArgumentsCount())
+                .mapToObj(i -> pF.convert(declaration, i, windowFunction.getArguments(i)))
+                .collect(java.util.stream.Collectors.toList());
+        var partitionExprs =
+            windowFunction.getPartitionsList().stream()
+                .map(this::from)
+                .collect(java.util.stream.Collectors.toList());
+        var sortFields =
+            windowFunction.getSortsList().stream()
+                .map(
+                    s ->
+                        Expression.SortField.builder()
+                            .direction(Expression.SortDirection.fromProto(s.getDirection()))
+                            .expr(from(s.getExpr()))
+                            .build())
+                .collect(java.util.stream.Collectors.toList());
+        var wfi =
+            WindowFunctionInvocation.builder()
+                .addAllArguments(args)
+                .declaration(declaration)
+                .outputType(protoTypeConverter.from(windowFunction.getOutputType()))
+                .aggregationPhase(Expression.AggregationPhase.fromProto(windowFunction.getPhase()))
+                .addAllSort(sortFields)
+                .invocation(
+                    Expression.AggregationInvocation.fromProto(windowFunction.getInvocation()))
+                .build();
+
+        WindowBound lowerBound = toLowerBound(windowFunction.getLowerBound());
+        WindowBound upperBound = toUpperBound(windowFunction.getUpperBound());
+
+        var wf = ImmutableExpression.WindowFunction.builder().function(wfi).build();
+        yield Expression.Window.builder()
+            .windowFunction(wf)
+            .hasNormalAggregateFunction(false)
+            .type(protoTypeConverter.from(windowFunction.getOutputType()))
+            .partitionBy(partitionExprs)
+            .orderBy(sortFields)
+            .lowerBound(lowerBound)
+            .upperBound(upperBound)
             .build();
```

Comment:
The fact that `WindowFunctionInvocation` and `Window` both share so many of the same fields seem like a bit of a code smell. I think it's possible to reduce the duplication, but decided to leave that off for a different change.